### DISCUSSION TITLE
Testing protected routes

### DISCUSSION THREAD

**Discussion by alexferrari88** [fastapi-users/fastapi-users#602] [Q&A] [answered]
Hello everyone. 

I created some routes that require a user to be logged in. 
For instance:
```python
@app.post(
    "/alerts",
    response_model=schemas.Alert,
    tags=["alerts"],
)
async def create_alert(
    alert: schemas.AlertCreate, owner: models.User = Depends(current_user)
):
    return await crud.create_alert(alert=alert, owner=owner)
```

I'm very new to testing and pytest and wanted to write some tests for these routes but I'm a little bit clueless on how to approach the whole thing. I have spent already countless hours researching and reading documentation online but I think I just need a little nudge in the right direction.

This is what I'm thinking:

1. In conftest.py I should create a user in my test database (I'm using Tortoise)
2. Get a test client for my app using httpx and LifespanManager (like in the tests for this very plugin)

The question is how can I then have this user logging in?

Since I cannot find code for testing this example anywhere, I would appreciate if you could share some example code with me.

Thank you!

**Comment by frankie567** [comment, marked as answer]
Hi @alexferrari88!

Currently, there is no testing utility to help you automating the authentication of a user in the test client.

What I do in my own projects is to manually generate a token from the authentication class and set the headers on the test client. It looks like this (I use the HTTPX/lifespan-manager approach for pure async tests):

```py
@pytest.fixture
def get_test_client():
    async def _get_test_client(user: UserDB = None):
        async with LifespanManager(app):
            async with httpx.AsyncClient(
                app=app, base_url="http://app.io"
            ) as test_client:
                if user:
                    jwt = await cookie_authentication._generate_token(user)
                    test_client.headers[
                        "Cookie"
                    ] = f"{cookie_authentication.cookie_name}={jwt}"
                yield test_client

    return _get_test_client
```

This is a pytest fixture factory that accepts a user as parameter. What I do then is to have some variations for unauthenticated/authenticated fixtures at hand:

```py
@pytest.fixture
@pytest.mark.asyncio
async def test_client(get_test_client):
    async for client in get_test_client():
        yield client

@pytest.fixture
@pytest.mark.asyncio
async def test_client_regular_user(get_test_client, regular_user):
    async for client in get_test_client(regular_user):
        yield client
```

You can then use `test_client` or `test_client_regular_user` directly in your tests:

```py
async def test_logged_out(self, test_client):
    response = await test_client.get("/api/affiliations/")
    assert response.status_code == status.HTTP_401_UNAUTHORIZED
```

Admittedly, this is a bit tedious and FastAPI Users could help developers by providing test utilities to help with this (like Django does). I'm thinking about it!

**Reply by alexferrari88** [reply]
thank you @frankie567, that is extremely useful!
just to clarify this for myself:
1. where do you get `cookie_authentication` from?
2. when you pass `regular_user` to the function, I suppose that is a fixture. Is this a fake user that you create with `UserDB(email, etc...)`?

**Reply by frankie567** [reply]
1. `cookie_authentication` is the FastAPI Users authentication class, as explained in the configuration (can be CookieAuthentication or JWTAuthentication): https://frankie567.github.io/fastapi-users/configuration/authentication/jwt/ or https://frankie567.github.io/fastapi-users/configuration/authentication/cookie/
2. Yes, it's a fixture returning a UserDB; you should create it in the database to make it work (FastAPI Users always check the Cookie/JWT against the database while authenticating).

**Reply by frankie567** [reply]
Thank you so much for the sponsorship @alexferrari88 🙏🎉 Much appreciated!

**Reply by alexferrari88** [reply]
@frankie567 I'm learning a lot by reading your code and tests, so **thank you** for this great work. I wish I could contribute more, hopefully in the future!

**Reply by m-danya** [reply]
Thanks, @frankie567 ! It would be great if this code snippet, along with [this one](https://github.com/fastapi-users/fastapi-users/discussions/830#discussioncomment-1874201), were included in the documentation. It really fits into the "Cookbook" section and could save newcomers some time. Much appreciated!

**Comment by artix-paladin** [comment]
Hello @alexferrari88 I was wondering how did you manage to protect your routes and only authorize it if there's a logged-in user. I am following the steps in the docs but It didn't work for me. Any help is appreciated. Thank you

**Reply by frankie567** [reply]
@hence-saber You should start your own question and provide details on what you did and what isn't working so that we can help you.

**Reply by artix-paladin** [reply]
Got it @frankie567 thank you
